### ISSUE TITLE
Implement Scrubbing

### ISSUE THREAD

**Issue by abrichr** [MLDSAI/OpenAdapt#47]
We would like to make it easy to scrub private information (e.g. PII/PHI):

```
python -m openadapt.scrub
```

**Comment by KrishPatel13** [comment]
Dear @abrichr,

I would like to work on implementing this issue. Just to confirm, my understanding is that we need to scrub private information from the input text such as:
- Date of Births
- Phone Numbers
- Bank Card Numbers
- Email Addresses

Please let me know if there are any additional requirements or specifications that I should consider while implementing the solution.

**Comment by abrichr** [comment]
Excellent, thank you Krish! These are the top Google results for [pii python], this looks promising, please start here:

https://pypi.org/project/piicatcher/
https://pypi.org/project/pii-extract/

Here is a Phind.com search that may be insructive:

https://www.phind.com/search?q=how%20to%20implement%20a%20PII%2FPHI%20scrubber%20in%20python%20for%20text%20and%20images%3F%20we%20already%20have%20OCR%20available&c=&source=searchbox

Please provide a few options and tradeoffs for each.

💪 let's do this 💪

**Comment by abrichr** [comment]
Phind.com recommends https://github.com/microsoft/presidio. After a glance it seems the top candidate (please let me know if you identify any other candidates worth comparing).

In your analysis please include a short description of the available APIs, e.g. function signatures, data sources/formats.

Thank you!!

**Comment by KrishPatel13** [comment]
Here's my analysis of the three options you've mentioned for implementing a PII/PHI scrubber in Python, along with their tradeoffs:

### **piicatcher:**

- piicatcher is a Python package for detecting personally identifiable information (PII) in text using regular expressions.
- It currently supports the detection of email addresses, credit card numbers, US phone numbers, US Social Security numbers, and IP addresses.
- The package is easy to install and use, and it provides a simple API for detecting PII in text.
- Tradeoff: it only supports a limited number of PII types, and it relies solely on regular expressions for detection, which can result in false positives or missed detections.
- Sample Usage API:
    • Function Signature: 
        `def catch(text: str, include_pattern_names: bool = False) -> List[str]:`
    • The function takes in a string of text as input and returns a list of PII entities detected in the text.


### **pii-extract:**

- pii-extract is a Python package for extracting PII from text using machine learning models.
- It supports the extraction of various types of PII, including email addresses, phone numbers, credit card numbers, and dates of birth.
- The package is easy to install and use, and it provides a simple API for extracting PII from text.
- Tradeoff: it requires a training set of labeled data to build the machine learning models, and the quality of the extraction depends on the quality of the training data.
- Sample Usage API:
    • Function Signature: 
        `def extract_entities(text: str) -> List[Dict[str, str]]:`
    • The function takes a string of text as input and returns a list of PII entities extracted from the text.
    • The output is a list of dictionaries, where each dictionary represents a single PII entity and contains the following fields: entity_type, entity_value, start, and end.
    • The entity_type field represents the type of PII entity detected (e.g., "EMAIL_ADDRESS", "PHONE_NUMBER", etc.), and the entity_value field represents the value of the entity detected.


### **Presidio:**

- Presidio is an open-source Python package for detecting and anonymizing PII in text and images using machine learning models.
- It supports the detection and anonymization of various types of PII, including names, addresses, phone numbers, email addresses, credit card numbers, and social security numbers.
- The package provides a comprehensive set of APIs for detecting and anonymizing PII, and it can be easily integrated into existing workflows.
- Tradeoff: it requires a training set of labeled data to build the machine learning models, and the quality of the detection and anonymization depends on the quality of the training data.
- Sample Usage APIs:
    • `analyze_text`: a function that takes a string of text as input and returns a list of PII entities detected in the text. The function has the following signature: 
    `def analyze_text(text: str, language: str = "en", score_threshold: float = 0.5) -> List[Entity]`
    • `anonymize_text`: a function that takes a string of text and a list of PII entities as input and returns the text with the PII entities anonymized. The function has the following signature: 
    `def anonymize_text(text: str, entities: List[Entity], replacement_str: str = "<PII>") -> str`
    • `analyze_image`: a function that takes an image as input and returns a list of PII entities detected in the image. The function has the following signature: 
    `def analyze_image(image: Union[bytes, str, IO]) -> List[Entity]`
    • `anonymize_image`: a function that takes an image and a list of PII entities as input and returns the image with the PII entities anonymized. The function has the following signature: 
    `def anonymize_image(image: Union[bytes, str, IO], entities: List[Entity], replacement_color: Tuple[int, int, int] = (0, 0, 0)) -> bytes`
    • The `Entity` class represents a single PII entity and contains the following fields: entity_type, start, end, and score.
    • The `entity_type` field represents the type of PII entity detected, and the start and end fields represent the start and end indices of the entity in the input text or image.
    • The `score` field represents the confidence score of the detection.
    
    
Overall, I agree with your view point, that is, we should use "`Presidio`".
**My Reasoning:** 
- It provides a wide range of APIs for detecting and anonymizing PII/PHI in both text and images, and supports a variety of languages and entity types.

- Presidio also offers advanced features such as custom entity recognition patterns, contextual analysis, and entity linking, which can help improve the accuracy and coverage of PII detection.

**Comment by KrishPatel13** [comment]
> Please also consider where in the existing PuterBot code this should be implemented. Some options:
> - before recording
> - before visualizing
> - before modelling/replay
> - before sharing


### **Analysis:**

- before recording: 
It would not benefit to Scrub before Recording, as we will have the ocr text generated only after the recording. Hence, we can rule out this option.

- before visualizing: 
In my opinion, we should Scrub the text before visualizing as it will help to visualize the data without exposing the sensitive information.

- before modelling/replay: 
In my opinion, we should Scrub the text before modeling/replay as it will help to model/replay the data without exposing the sensitive information.

- before sharing: 
In my opinion, implementing scrubbing before sharing is the most critical step for protecting privacy and ensuring compliance with data protection regulations.


BUT, scrubbing at multiple stage may be time consuming. I would recommend implementing scrubbing at the earliest possible stage where it can be effectively applied ONCE, and then it may be reused by any other stage/requirement.

**Comment by abrichr** [comment]
I agree that we should scrub at the earliest possible stage. How about at the top of [process_event](https://github.com/MLDSAI/puterbot/blob/main/puterbot/record.py#L60)?

**Comment by KrishPatel13** [comment]
I agree, we can provide with this idea. BUT,
We would need the text for scrubbing. Will it be something like,

text = event.text #Accessing the InputEvent's ._text attribute
scubbed_text = scrub(text)
# Then, we need to set the new scrubbed text back to the event's _text attribute.
# So, we need to add a new setter method in models.py under the InputEvent class.

Please correct me if my approach is incorrect.

**Comment by abrichr** [comment]
Hi @KrishPatel13 , on further consideration, there are different types of scrubbing we need to implement depending on the event:

- [write_screeen_event](https://github.com/MLDSAI/puterbot/blob/main/puterbot/record.py#L183) should scrub image data
- [write_window_event](https://github.com/MLDSAI/puterbot/blob/main/puterbot/record.py#L189) should scrub text data from window titles and window state (work in progress, https://github.com/MLDSAI/puterbot/issues/75)
- [write_input_event](https://github.com/MLDSAI/puterbot/blob/main/puterbot/record.py#L148) is trickier because a single input event may not contain PII (e.g. the letter "b"), but a sequence of them will (e.g. the sequence "bob smith") 

InputEvents are merged in [get_consecutive_keyboard_events](https://github.com/MLDSAI/puterbot/blob/main/puterbot/events.py#L389). This logic is nontrivial and currently incomplete, and may run several times before everything is merged.

The merging process is completed after the `process`  loop is completed in [get_events](https://github.com/MLDSAI/puterbot/blob/main/puterbot/events.py#L69). This is probably the better place to implement scrubbing of InputEvents.

Eventually we will want to save the processed events to new tables, which are ultimately what will be published (https://github.com/MLDSAI/puterbot/issues/46). This creation of new tables has not been implemented yet, and ideally should be implemented as part of this work, perhaps as a separate PR.

**Comment by KrishPatel13** [comment]
@abrichr 
[write_screeen_event](https://github.com/MLDSAI/puterbot/blob/main/puterbot/record.py#L183) should scrub image data

I am a bit unsure for the approach here. This is because, the given a binary string of png_data. If we want to scrub it,

We will need to decode using base64 and then scrub the text. Then again encoding to the png_data.

Please let me know if this approach is correct?

**Comment by KrishPatel13** [comment]
My Progress: 
Before:
![image](https://github.com/MLDSAI/OpenAdapt/assets/65433817/5080d8d1-5478-4638-8b93-ac44f2de016a)

After:
![image](https://github.com/MLDSAI/OpenAdapt/assets/65433817/4150ce7c-b112-40f3-94a3-93657fd72967)

**Comment by KrishPatel13** [comment]
@abrichr 

Ready for Review! [My PR](https://github.com/MLDSAI/OpenAdapt/pull/128)
